src/main/java/randoop/generation/constanttfidf/ScopeStatistics.java

```diff
@@ -0,0 +1,118 @@
+package randoop.generation.constanttfidf;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import randoop.main.GenInputsAbstract;
+import randoop.sequence.Sequence;
+
+/**
+ * This class stores the constant mining information. It stores the frequency of the sequence, the
+ * number of classes that contain the sequence, and the total number of classes in the current
+ * scope.
+ */
+public class ScopeStatistics {
+
+  // These two fields have the same keyset.
+
+  /** A map from a constant to the number of times it is used in the current scope. */
+  Map<Sequence, Integer> numUses;
+
+  /**
+   * A map from a constant to the number of classes in the current scope that contains it. Null if
+   * the literals level is CLASS.
+   */
+  @Nullable Map<Sequence, Integer> numClassesWith;
+
+  /** The number of classes in the given scope. */
+  int numClasses;
+
+  /** Creates a new empty ScopeStatistics. */
+  public ScopeStatistics() {
+    numUses = new HashMap<>();
+    switch (GenInputsAbstract.literals_level) {
+      case CLASS:
+        // Since CLASS level regards the class that the constant locate as its scope, no need to
+        // store the classesWithConstant and numClasses.
+        numClassesWith = null;
+        numClasses = 1;
+        break;
+      case PACKAGE:
+      case ALL:
+        // Since the ALL level uses the whole project as its scope, the ALL_SCOPE key is used to
+        // store the classesWithConstant and numClasses.
+        numClassesWith = new HashMap<>();
+        numClasses = 0;
+        break;
+      default:
+        throw new RuntimeException("Unknown literals level");
+    }
+  }
+
+  /**
+   * Get the frequency information.
+   *
+   * @return the frequency information
+   */
+  public Map<Sequence, Integer> getNumUses() {
+    return numUses;
+  }
+
+  /**
+   * Get the classesWithConstant information.
+   *
+   * @return the classesWithConstant information
+   */
+  public @Nullable Map<Sequence, Integer> getNumClassesWith() {
```

Comment:
This return type is `@Nullable Map<...>`.  However, the documentation does not indicate that the result may be null.  I think that the return value is never null during normal, expected execution.  If that is the case, then `@Nullable` should not be written here.  Only write `@Nullable` if `null` can be returned by a normal execution, and call sites need to check for the null value.